File: Examples/CSharp/x-IMU3-Examples/Examples/PortScanner.cs
namespace Ximu3Examples
{
    class PortScanner
    {
        public PortScanner()
        {
            // Blocking
            var devices = Ximu3.PortScanner.Scan();

            PrintDevices(devices);

            // Non-blocking
            using var portScanner = new Ximu3.PortScanner(Callback);

            System.Threading.Thread.Sleep(60000);
        }

        private void Callback(Ximu3.Device[] devices)
        {
            PrintDevices(devices);
        }

        private static void PrintDevices(Ximu3.Device[] devices)
        {
            Console.WriteLine(devices.Length + " device(s) found");

            foreach (var device in devices)
            {
                Console.WriteLine(
                    Ximu3.Helpers.ToString(device.wrapped.device_name) + ", " +
                    Ximu3.Helpers.ToString(device.wrapped.serial_number) + ", " +
                    Ximu3.ConnectionConfig.From(device)
                );
                // Console.WriteLine(device.ToString()); // alternative to above
            }
        }
    }
}

File: Examples/Python/port_scanner.py
import time

import ximu3


# Blocking
def print_devices(devices: list[ximu3.Device]) -> None:
    print(f"{len(devices)} device(s) found")

    for device in devices:
        print(", ".join([device.device_name, device.serial_number, str(device.connection_config)]))
        # print(device)  # alternative to above


devices = ximu3.PortScanner.scan()

print_devices(devices)


# Non-blocking
def callback(devices: list[ximu3.Device]) -> None:
    print_devices(devices)


port_scanner = ximu3.PortScanner()

port_scanner.add_callback(callback)

time.sleep(60)
